app_dart/lib/src/request_handlers/task_status_subscription.dart

```diff
@@ -0,0 +1,127 @@
+///
+/// The PubSub subscription is set up here:
+/// https://cloud.google.com/cloudpubsub/subscription/detail/luci-builds?project=flutter-dashboard&tab=overview
+///
+/// This endpoint is responsible for updating Datastore with the result of tasks from LUCI.
+@immutable
+class TaskStatusSubscription extends SubscriptionHandler {
+  const TaskStatusSubscription(
+    CacheService cache,
+    Config config,
+    AuthenticationProvider authProvider, {
+    @visibleForTesting this.datastoreProvider = DatastoreService.defaultProvider,
+  }) : super(cache: cache, config: config, authProvider: authProvider, topicName: 'luci-builds');
+
+  final DatastoreServiceProvider datastoreProvider;
+
+  @override
+  Future<Body> post() async {
+    final String data = message.data!;
+    final BuildPushMessage buildPushMessage =
+        BuildPushMessage.fromJson(json.decode(String.fromCharCodes(base64.decode(data))) as Map<String, dynamic>);
+
+    final Map<String, dynamic> userData = jsonDecode(buildPushMessage.userData!) as Map<String, dynamic>;
+    final String? commitBranch = userData['commit_branch'] as String?;
+    if (commitBranch == null) {
+      throw const BadRequestException('userData does not contain commit_branch');
+    }
+    final String? commitSha = userData['commit_sha'] as String?;
+    if (commitSha == null) {
+      throw const BadRequestException('userData does not contain commit_sha');
+    }
+    final String? builderName = userData['builder_name'] as String?;
+    if (builderName == null) {
+      throw const BadRequestException('userData does not contain builder_name');
+    }
+    final String? newStatus = buildPushMessage.build?.status as String?;
+    if (newStatus == null) {
+      throw const BadRequestException('userData does not contain builder_name');
+    }
+
+    final DatastoreService datastore = datastoreProvider(config.db);
+
+    if (newStatus != Task.statusSucceeded && newStatus != Task.statusFailed) {
+      throw const BadRequestException('NewStatus can be one of "Succeeded", "Failed"');
+    }
+
+    final Task task = await _getTaskFromNamedParams(datastore, builderName, commitBranch, commitSha);
+
+    task.status = newStatus;
+    task.endTimestamp = DateTime.now().millisecondsSinceEpoch;
+
+    await datastore.insert(<Task>[task]);
+    return Body.empty;
+  }
+
+  /// Retrieve [Task] from [DatastoreService] when given [commitSha], [commitBranch], and [builderName].
+  ///
+  /// This is used when the DeviceLab test runner is uploading results to Cocoon for runs on LUCI.
+  /// LUCI does not know the [Key] assigned to task when scheduling the build, but Cocoon can
+  /// lookup the task based on these key values.
+  ///
+  /// To lookup the value, we construct the ancestor key, which corresponds to the [Commit].
+  /// Then we query the tasks with that ancestor key and search for the one that matches the builder name.
+  Future<Task> _getTaskFromNamedParams(
```

Comment:
This logic should be moved into a dedicated results service so it can be shared with `/api/update-task-status`

Reply:
Ideally we would deprecate `/api/update-task-status` as soon as we can. So we should not need to live with the duplicated code for too long. Do you still want it broken out for now?

Reply:
\cc @renyou - WDYT the work would be to move FRoB from using the REST API to publishing messages to PubSub? We're migrating that logic from HTTP to PubSub, but we can maintain it for now

Reply:
> oB
I believe that's a question for you/xilai and me as we are becoming the owners of FRoB starting next week. Maybe a good start project to start ramping up on FRoB?